libclamav/ole2_extract.c

```diff
@@ -1756,6 +2024,423 @@ ole2_read_header(int fd, ole2_header_t *hdr)
+#define SE_HEADER_EI_AES256 0x00006610
+
+#define SE_HEADER_EI_RC4 0x00006801
+
+    switch (headerPtr->encryptionInfo.algorithmID) {
+        case SE_HEADER_EI_AES128:
+            break;
+        case SE_HEADER_EI_AES192:
+            //not implemented
+            goto done;
+        case SE_HEADER_EI_AES256:
+            //not implemented
+            goto done;
+        case SE_HEADER_EI_RC4:
+            //not implemented
+            goto done;
+        default:
+            cli_dbgmsg( "ole2: Invalid Algorithm ID: 0x%x\n", headerPtr->encryptionInfo.algorithmID);
+            goto done;
+    }
+
+#define SE_HEADER_EI_SHA1 0x00008004
+    if (SE_HEADER_EI_SHA1 != headerPtr->encryptionInfo.algorithmIDHash) {
+        cli_warnmsg("ole2: Invalid Algorithm ID Hash: 0x%x\n", headerPtr->encryptionInfo.algorithmIDHash);
+        goto done;
+    }
+
+//#define SE_HEADER_EI_AES128_KEYSIZE 0x00000080
+//#define SE_HEADER_EI_AES192_KEYSIZE 0x000000c0
+//#define SE_HEADER_EI_AES256_KEYSIZE 0x00000100
+
+    switch (headerPtr->encryptionInfo.keySize) {
+        case SE_HEADER_EI_AES128_KEYSIZE:
+            break;
+        case SE_HEADER_EI_AES192_KEYSIZE:
+            break;
+        case SE_HEADER_EI_AES256_KEYSIZE:
+            break;
+        default:
+            cli_warnmsg("ole2: Invalid key size: 0x%x\n", headerPtr->encryptionInfo.keySize);
+            goto done;
+    }
+    cli_dbgmsg( "KeySize = 0x%x\n", headerPtr->encryptionInfo.keySize);
+
+#define SE_HEADER_EI_AES_PROVIDERTYPE 0x00000018
+    if (SE_HEADER_EI_AES_PROVIDERTYPE != headerPtr->encryptionInfo.providerType) {
+        cli_warnmsg("ole2: WARNING: Provider Type should be '0x%x', is '0x%x'\n",
+                    SE_HEADER_EI_AES_PROVIDERTYPE, headerPtr->encryptionInfo.providerType);
+        goto done;
+    }
+
+    cli_dbgmsg( "Reserved1:  0x%x\n", headerPtr->encryptionInfo.reserved1);
+
+    if (0 != headerPtr->encryptionInfo.reserved2) {
+        cli_warnmsg("ole2: Reserved 2 must be zero, is 0x%x\n", headerPtr->encryptionInfo.reserved2);
+        goto done;
+    }
+
+    /*The encryption info is at the end of the CPSName string.  
+     * Find the end, and we'll have the index of the EncryptionVerifier.
+     * The CPSName string *should* always be either
+     * 'Microsoft Enhanced RSA and AES Cryptographic Provider'
+     * or
+     * 'Microsoft Enhanced RSA and AES Cryptographic Provider (Prototype)'
+     *
+     */
+    for (idx = 0; idx < sizeof(headerPtr->encryptionInfo.cspName) - 1; idx += 2) {
+        if (((uint16_t *)&(headerPtr->encryptionInfo.cspName[idx]))[0] == 0) {
+            break;
+        }
+    }
+
+    idx += 2;
+    if ((sizeof(headerPtr->encryptionInfo.cspName) - idx) <= sizeof(encryption_verifier_t)) {
+        cli_warnmsg("ole2: ERROR: No encryption_verifier_t\n");
+        goto done;
+    }
+    copy_encryption_verifier(&ev, &(headerPtr->encryptionInfo.cspName[idx]));
+
+    key.key_len = headerPtr->encryptionInfo.keySize / 8;
```

Comment:
Why are we storing `keySize / 8` in the `key.key_len` and then multiplying it by `8` everywhere it is used?  Can we just store `keySize` directly?